test/spec/modules/conversantAnalyticsAdapter_spec.js

```diff
@@ -0,0 +1,733 @@
+        bidderCode: 'bidderCode',
+        auctionId: 'auctionId',
+        timeReceived: Date.now()
+      };
+      expect(requests.length).to.equal(0);
+      expect(Object.keys(caObj.adIdLookup).length).to.equal(1);
+      events.emit(constants.EVENTS.AD_RENDER_FAILED, RENDER_FAILED_PAYLOAD);
+      expect(Object.keys(caObj.adIdLookup).length).to.equal(0); // object should be removed but no call made to send data
+      expect(requests.length).to.equal(0);
+    });
+  });
+
+  describe('Bid Won Tests', function() {
+    const GOOD_BID_WON_ARGS = {
+      bidderCode: 'conversant',
+      width: 300,
+      height: 250,
+      statusMessage: 'Bid available',
+      adId: '57e03aeafd83a68',
+      requestId: '2c2a5485a076898',
+      mediaType: 'banner',
+      source: 'client',
+      currency: 'USD',
+      cpm: 4,
+      creativeId: '29123_55016759',
+      ttl: 300,
+      netRevenue: true,
+      ad: '<foobar add goes here />',
+      originalCpm: 0.04,
+      originalCurrency: 'USD',
+      auctionId: '85e1bf44-4035-4e24-bd3c-b1ba367fe294',
+      responseTimestamp: 1583851418626,
+      requestTimestamp: 1583851418292,
+      bidder: 'conversant',
+      adUnitCode: 'div-gpt-ad-1460505748561-0',
+      timeToRespond: 334,
+      pbLg: '4.00',
+      pbMg: '4.00',
+      pbHg: '4.00',
+      pbAg: '4.00',
+      pbDg: '4.00',
+      pbCg: '',
+      size: '300x250',
+      adserverTargeting: {
+        hb_bidder: 'conversant',
+        hb_adid: '57e03aeafd83a68',
+        hb_pb: '4.00',
+        hb_size: '300x250',
+        hb_source: 'client',
+        hb_format: 'banner'
+      },
+      status: 'rendered',
+      params: [
+        {
+          site_id: '108060'
+        }
+      ]
+    };
+
+    // no adUnitCode, auctionId or bidderCode will cause a failure
+    const BAD_BID_WON_ARGS = {
+      bidderCode: 'conversant',
+      width: 300,
+      height: 250,
+      statusMessage: 'Bid available',
+      adId: '57e03aeafd83a68',
+      requestId: '2c2a5485a076898',
+      mediaType: 'banner',
+      source: 'client',
+      currency: 'USD',
+      cpm: 4,
+      originalCpm: 0.04,
+      originalCurrency: 'USD',
+      bidder: 'conversant',
+      adUnitCode: 'div-gpt-ad-1460505748561-0',
+      size: '300x250',
+      status: 'rendered',
+      params: [
+        {
+          site_id: '108060'
+        }
+      ]
+    };
+
+    beforeEach(function () {
+      conversantAnalytics.enableAnalytics(VALID_ALWAYS_SAMPLE_CONFIG);
+    });
+
+    afterEach(function () {
+      conversantAnalytics.disableAnalytics();
+    });
+
+    it('should not send data or put a record in adIdLookup when bad data provided', function() {
+      expect(requests.length).to.equal(0);
+      expect(Object.keys(caObj.adIdLookup).length).to.equal(0);
+      events.emit(constants.EVENTS.BID_WON, BAD_BID_WON_ARGS);
+      expect(requests.length).to.equal(0);
+      expect(Object.keys(caObj.adIdLookup).length).to.equal(0);
```

Comment:
If somehow prebid is not behaving, we these errors should be collected and reported.  Probably something for future.  Maybe not having adIdLookup cache is ok.  Why not just send back whatever is available so at least the data is there for post mortem.

Reply:
The adId lookup is there so that we can match it up to the offending bid_won potentially... otherwise the information we get is useless if you look at what is actually provided to us in a RENDER_FAILED payload.  I agree that perhaps having some cache to store any bad/unexpected requests so we know if something changed in Prebid we didn't anticipate that we get this information rather than just tossing it.  I'll add a story for this for next phase stuff I guess